sigma/plugins/roles/setcolor.py

```diff
@@ -0,0 +1,63 @@
+import discord
+from sigma.core.permission import check_man_roles
+from sigma.core.rolecheck import matching_role
+
+async def setcolor(cmd, message, args):
+    # if no args, send help message
+    if not args:
+        await message.channel.send(cmd.help())
+        return
+
+    # checks permissions
+    if not check_man_roles(message.author, message.channel):
+        out_content = discord.Embed(type='rich', color=0xDB0000, title='⛔ Error')
+        out_content.add_field(name='Insufficient Permissions', value='You require the Manage Roles permission.')
+        await message.channel.send(None, embed=out_content)
+        return
+
+    color = args[0].lstrip('#').upper()
+    # check if color is a proper format
+    if not is_valid_hex(color):
+        out_content = discord.Embed(type='rich', color=0xFF9900, title='⚠ Error')
+        out_content.add_field(name='Invalid Input', value='A valid hexadecimal color code has numbers from 0–9 and letters from A–F, and has six characters.')
+        await message.channel.send(None, embed=out_content)
+        return
+
+    # check if role exists
+    exists = matching_role(message.guild, color)
+    if not exists:
+        discord_colour = discord.Colour(int(color, 16))
+        target_role = await message.guild.create_role(name=color, colour=discord_colour, reason="Color role")
+    else:
+        target_role = exists
+
+    target = message.author if not message.mentions else message.mentions[0]
+
+    # get user's roles, remove old role, append new role, and set user's roles
+    list_of_roles = target.roles
+
+    for role in list_of_roles:
+        if is_valid_hex(role.name):
+            if len(role.members) == 1:
+                await role.delete()
+            else:
+                await target.remove_roles(role)
```

Comment:
Issues with this:
* This part runs regardless whether the user already has color role
* This part checks if only one person has the role. What if it's more than 1?

Reply:
> This part checks if only one person has the role. What if it's more than 1?

If more than one person has the role, then it goes to the `else` block, which removes the role from the target user instead of nuking the role.

Reply:
> This part runs regardless whether the user already has color role

Do you want the command to stop early on if the user already has the role?

Reply:
> If more than one person has the role, then it goes to the else block, which removes the role from the target user instead of nuking the role.

Oh ok. Is it possible that you get a 0?


> Do you want the command to stop early on if the user already has the role?

Yes

Reply:
> Is it possible that you get a 0?
No. This loop loops through the roles the target user has. You can't get a zero, because then the role wouldn't show up in the list of roles, and it wouldn't be looped through.

Reply:
ok

Reply:
Done.